### PR TITLE
[DLP] Implemented Inspect Table With Custom Hotword

### PR DESCRIPTION
**Pull request by bhavika-dhanwani**
Implemented [Inspect Table With Custom Hotword ](https://cloud.google.com/dlp/docs/creating-custom-infotypes-likelihood.md#match-column-values)

## Checklist

- [X] I have followed [Sample Format Guide](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md)
- [X] `pom.xml` parent set to latest `shared-configuration`
- [ ] Appropriate changes to README are included in PR
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [X] **Tests** pass:   `mvn clean verify` **required**
- [X] **Lint**  passes: `mvn -P lint checkstyle:check` **required**
- [ ] **Static Analysis**:  `mvn -P lint clean compile pmd:cpd-check spotbugs:check` **advisory only**
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample 
- [X] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: dlp/snippets/src/main/java/dlp/snippets/InspectTableWithCustomHotword.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                    .addValues(Value.newBuilder().setStringValue("111-11-1111").build())
+                    .addValues(Value.newBuilder().setStringValue("222-22-2222").build())
+                    .build())
+            .build();
+    // Specify the regex pattern to be detected.
+    String hotwordRegexPattern = "(Fake Social Security Number)";
+    inspectWithCustomHotwords(projectId, tableToDeIdentify, hotwordRegexPattern);
+  }
+
+  //  Sets the match likelihood of an entire column of data.
```
Comment (Sita04) [inline dlp/snippets/src/main/java/dlp/snippets/InspectTableWithCustomHotword.java:64]: The description is not clear. Kindly brief it.

File: dlp/snippets/src/test/java/dlp/snippets/InspectTests.java [hunk 2]
```diff
@@ -416,4 +416,25 @@ public void testInspectStringAugmentInfoType() throws Exception {
...
+            .addHeaders(FieldId.newBuilder().setName("Fake Social Security Number").build())
+            .addHeaders(FieldId.newBuilder().setName("Real Social Security Number").build())
+            .addRows(
+                Table.Row.newBuilder()
+                    .addValues(Value.newBuilder().setStringValue("111-11-1111").build())
+                    .addValues(Value.newBuilder().setStringValue("222-22-2222").build())
+                    .build())
+            .build();
+    InspectTableWithCustomHotword.inspectWithCustomHotwords(
+        PROJECT_ID, tableToDeIdentify, "(Fake Social Security Number)");
```
[the remark this answers is not in the document]
Comment (bhavika-dhanwani) [reply dlp/snippets/src/test/java/dlp/snippets/InspectTests.java:433]: @Sita04 Thank you for reviewing the sample. According to me, the provided functionality excludes the results of any finding that matches the US_SOCIAL_SECURITY_NUMBER infoType detector from an entire column of tabular data, provided the column name matches a given regular expression. Specifically, in the given example, any such finding within the `Fake Social Security Number` column is excluded from the results. So, we can't proceed with this change, please let me know if I'm missing anything from my side.

File: dlp/snippets/src/main/java/dlp/snippets/InspectTableWithCustomHotword.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            .addHeaders(FieldId.newBuilder().setName("Fake Social Security Number").build())
+            .addHeaders(FieldId.newBuilder().setName("Real Social Security Number").build())
+            .addRows(
+                Table.Row.newBuilder()
+                    .addValues(Value.newBuilder().setStringValue("111-11-1111").build())
+                    .addValues(Value.newBuilder().setStringValue("222-22-2222").build())
+                    .build())
+            .build();
+    // Specify the regex pattern to be detected.
+    String hotwordRegexPattern = "(Fake Social Security Number)";
```
Comment (micheldavid) [inline dlp/snippets/src/main/java/dlp/snippets/InspectTableWithCustomHotword.java:59]: no need to put text in parenthesis. might want to add a link to RE2 regex syntax page: https://github.com/google/re2/wiki/Syntax

File: dlp/snippets/src/main/java/dlp/snippets/InspectTableWithCustomHotword.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    inspectWithCustomHotwords(projectId, tableToDeIdentify, hotwordRegexPattern);
+  }
+
+  //  Inspects the provided table, excluding the findings of entire column matching regular
+  // expression.
+  public static void inspectWithCustomHotwords(
```
Comment (micheldavid) [inline dlp/snippets/src/main/java/dlp/snippets/InspectTableWithCustomHotword.java:65]: name suggestion: inspectDemotingFindingsWithHotwords
makes it more clear that we are changing finding likelihood to very unlikely when they are close to the hotword

File: dlp/snippets/src/main/java/dlp/snippets/InspectTableWithCustomHotword.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+public class InspectTableWithCustomHotword {
+
+  public static void main(String[] args) throws IOException {
+    // TODO(developer): Replace these variables before running the sample.
+    // The Google Cloud project id to use as a parent resource.
+    String projectId = "your-project-id";
+    // Specify the table to be considered for de-identification.
+    Table tableToDeIdentify =
+        Table.newBuilder()
+            .addHeaders(FieldId.newBuilder().setName("Fake Social Security Number").build())
```
Comment (micheldavid) [inline dlp/snippets/src/main/java/dlp/snippets/InspectTableWithCustomHotword.java:50]: i could see the engine improving to eventually exclude findings that are close to a word like "fake", "test", "invalid". can we make it a bit more obfuscated? suggestion: "Some Social Security Number"

File: dlp/snippets/src/test/java/dlp/snippets/InspectTests.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (micheldavid) [reply dlp/snippets/src/test/java/dlp/snippets/InspectTests.java:433]: i'd recommend against any use of randomization in tests.
static inputs provide consistent results, which is the goal for these kinds of tests.

File: dlp/snippets/src/test/java/dlp/snippets/InspectTests.java [hunk 6]
```diff
@@ -416,4 +416,25 @@ public void testInspectStringAugmentInfoType() throws Exception {
...
+
+    String output = bout.toString();
+    assertThat(output).contains("Findings: 1");
+    assertThat(output).contains("Info type: US_SOCIAL_SECURITY_NUMBER");
+    assertThat(output).contains("Likelihood: VERY_LIKELY");
```
Comment (micheldavid) [inline dlp/snippets/src/test/java/dlp/snippets/InspectTests.java:438]: good to confirm the finding here is the expected one: "222-22-2222"

**Review by micheldavid — APPROVED** [review]

**Review by Shabirmean — APPROVED** [review]

**Comment by bhavika-dhanwani** [comment]
@Shabirmean Could you please add [kokoro:force-run](https://github.com/GoogleCloudPlatform/java-docs-samples/labels/kokoro%3Aforce-run) label to run the required checks.

**Review by Sita04 — APPROVED** [review]

**Pull request merged**